Add release process reference guide

iameskild opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://nebari.dev/community
-->

## Reference Issues or PRs

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create a link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [x] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

## Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Nebari Internals docs.
  - If no docs exist:
    - Create a stub for documentation, including bullet points for how to use the feature, code snippets (including from
      happy path tests), etc.
We want to ensure that the content produced for Nebari is accessible; if you are making significant contributions,
please address the access-centred guidelines in your content and complete our checklist. Thanks to @isabela-pf for this checklist :)
-->

### Access-centered content checklist

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

## Any other comments?

<!--
Please be aware that we are a loose team of volunteers, so patience is necessary;
assistance handling other issues is very welcome.
We value all user contributions. If we are slow to review, either the pull request needs some benchmarking, tinkering,
convincing, etc., or the reviewers are likely busy. In either case,
we ask for your understanding during the
review process.
Thanks for contributing to Nebari 🙏🏼!
-->

iameskild commented:
@trallard @pavithraes I moved the release process doc to the community section and updated the Gitflow section to reflect what we discussed during the Nebari sync.

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+This page describes the high-level details of cutting a new Nebari release.
+
+|                   |                                                                                                                                                                                                  |
```
Need to remove this row - otherwise it renders as an empty row

iameskild replied:
I rotated the table instead 👍

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+|                   |                                                                                                                                                                                                  |
+| :---------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+|      Cadence      |                                                                                             Monthly                                                                                              |
+| Versioning System |                                                                                    [CalVer](#calver-details)                                                                                     |
```
I know this links to the section below but I'd still spell out the CalVer flavour we use - add an example

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+For every release, there is an assigned "Release Captain". It is the responsibility of this person to manage both the release and testing checklists as well as to communicate the status of the release to the rest of the Nebari development team and the community.
+
+Communication channels that need to be updated are:
+
+- the release checklist issue itself
+- the `#nebari-dev` slack channel
```
Since this is not a public channel and it's likely that folks will get confused I prefer removing this

trallard replied:
Thinking about this - the suggestion above should cover this
```suggestion
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+## Release Captain responsibilities
+
+For every release, there is an assigned "Release Captain". It is the responsibility of this person to manage both the release and testing checklists as well as to communicate the status of the release to the rest of the Nebari development team and the community.
```
```suggestion
For every release, there is an assigned "Release Captain". The Release Captain's responsibilities are:

- Manage both the release and testing processes and update the checklists as needed.
- Communicate the status of the release to the rest of the Nebari development team and the community (through updating the checklists and adding status updates as comments).
- Assign owners to checklist items (if not owned by the Release Captain).
- Adjust the schedule, particularly the publishing dates, based on defects found, fixes made, holidays, vacations, and so on.
- Ensure a new branch and milestone is created after completing a release cycle.
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+|                   |                                                                                                                                                                                                  |
+| :---------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+|      Cadence      |                                                                                             Monthly                                                                                              |
```
Monthly - on the fourth week of the month

iameskild replied:
I updated this with the `3rd week of the month`. This gives us a little buffer ;)

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+## CalVer details
+
+These releases should follow the following CalVer versioning style:
```
```suggestion
Nebari releases should follow the following CalVer versioning style:
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+For the release tag, there should be NO prepended `v`
+:::
+
+For example, the first Nebari CalVer release was `2022.10.1`. If a hotfix release was needed in the same month, we simply increment the `releaseNumber` by 1, which would be `2022.10.2` (_this is to illustrate how the increment works, this release does not exist._)
```
```suggestion
For example, the first Nebari CalVer release was `2022.10.1`. If a hotfix release was needed in the same month, we increment the `releaseNumber` by 1, which would be `2022.10.2` (_this is to illustrate how the increment works, this release does not exist._)
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+Gitflow is framework for managing `git` branches by assigning certain roles to particular branches.
+
+`main` - Represents a production-ready state of the code-base, with an appropriate tag to match the most recent release.
+
+`release/YYYY-MM-releas1eNumber` - Represents the branch for the upcoming release and is the _default_ branch on the GitHub repository.
```
```suggestion
- `main` - Represents a production-ready state of the code-base, with an appropriate tag to match the most recent release.

- `release/YYYY-MM-releas1eNumber` - Represents the branch for the upcoming release and is the _default_ branch on the GitHub repository.
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+Whenever a new release is out, it is the responsibility of the Release Captain to create a new release branch `release/YYYY-MM-releaseNumber` for the next month's release and to set this new branch as the default branch on the GitHub repository.
+
+All feature branches and bug fixes pull requests (PRs) will then be opened against this new release branch.
```
```suggestion
Active development occurs against the `release/YYYY-MM-releaseNumber` branch.
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+Although this process is captured in the [release checklist template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=release-checklist.md&title=%5BRELEASE%5D+%3Cversion%3E), it's worth making clear how branches are managed.
+
+Whenever a new release is out, it is the responsibility of the Release Captain to create a new release branch `release/YYYY-MM-releaseNumber` for the next month's release and to set this new branch as the default branch on the GitHub repository.
```
```suggestion
Whenever a new release is out, it is the responsibility of the Release Captain to create a new release branch, `release/YYYY-MM-releaseNumber` for the next month's release and to set this new branch as the default branch on the GitHub repository.
Setting this as the default branch ensures that new Pull Requests will automatically default to submitting against it. Force pushes are not allowed to these branches.
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+All feature branches and bug fixes pull requests (PRs) will then be opened against this new release branch.
+
+And the cycle continues.
```
```suggestion
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+And the cycle continues.
+
+:::note
+This process was inspired by the Bokeh [BEP6](https://github.com/bokeh/bokeh/wiki/BEP-6:-Branching-Strategy).
+:::
```
Suggest adding this as an acknowledgement section as we do for nebari.dev/docs/community/

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+---
+id: release-process
+title: The Nebari release process
```
```suggestion
title: Release process
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+---
+id: release-process
+title: The Nebari release process
+description: The Nebari release process and workflow
```
```suggestion
description: Release process
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,75 @@
+
+For example, the first Nebari CalVer release was `2022.10.1`. If a hotfix release was needed in the same month, we simply increment the `releaseNumber` by 1, which would be `2022.10.2` (_this is to illustrate how the increment works, this release does not exist._)
+
+## Gitflow and branching strategy
```
I know this is somewhat related to releases, but I would prefer having the gitflow + branching in a separate doc. 

As a newcomer, I would not look at the release process if I only care about how to open a PR and which branch to use.

iameskild replied:
As you mentioned, these concepts are closely linked and I believe they should remain together. We reference the "release process" quite a bit in the "Gitflow" section and if the our branching strategy makes sense, newcomers shouldn't need to decide which branch they need to open their PR against 😊

pavithraes replied:
I'd also like to make the branching strategy more visible/finable in the sidebar, for folks who might be looking exactly for this information. Bokeh contributors tend to look for these details when they notice the branch names because it is not a super common workflow.

If we decide to not create a separate doc, I think it'll be worth including it in the page title: "Release process and branching strategy". :)

trallard replied:
Much better

trallard commented:
Left several comments here, stuff that I think is missing:

- Neither the checklist nor these docs tell anyone how to create a tag. See https://numpy.org/devdocs/dev/releasing.html#prepare-the-release-commit 

From the checklist

```
- [ ] [Cut PyPI release via GHA release workflow.](https://github.com/nebari-dev/nebari/releases/new)
  - Avoid appending `v` to the tag.
    - Copy release notes from `RELEASE.md`.
```

- Also, where are the release notes copied to or from?

pavithraes commented:
> Also, where are the release notes copied to or from?

The actual release notes are recorded at: https://github.com/nebari-dev/nebari/blob/release/2022.11.1/RELEASE.md

We want to copy these to: https://www.nebari.dev/docs/references/RELEASE

I do have a PR open to copy these over automatically on each release: https://github.com/nebari-dev/nebari/pull/1554

iameskild commented:
Thanks for the review @trallard! I believe I addresses most of the concerns above. 

> * Neither the checklist nor these docs tell anyone how to create a tag. See https://numpy.org/devdocs/dev/releasing.html#prepare-the-release-commit
> 
> From the checklist
> 
> ```
> - [ ] [Cut PyPI release via GHA release workflow.](https://github.com/nebari-dev/nebari/releases/new)
>   - Avoid appending `v` to the tag.
>     - Copy release notes from `RELEASE.md`.
> ```

We don't need to manually create any tags as it is handled by GitHub's release process (which is where we link to) :)

pavithraes commented:
Thanks for working on this page, Eskild. I think it's looking great!

iameskild commented:
@trallard @pavithraes as I have been writing up these docs and trying to think through how (existing and new) contributors will likely work, I'm starting to have some reservations about the Bokeh style gitflow workflow. There are a few places where this gitflow workflow is not entirely intuitive and will likely cause some confusion. A few examples of this include: 
- not knowing which branch to branch off
  - users should branch off of the default branch `release/YYYY-MM-releaseNumber` but this will change every month
- if the feature you're are working on takes more than a month to complete, you'll need to rebase
  - if your target branch gets deleted (once its merged into `main`), does this also automatically close your PR?
- if for some reason there is no release during a particular month, then all open PRs will need to have their target branch updated

I'm not trying to drag this discussion out but I think it's worth making this process as simple and as intuitive as possible for the sake of all devs (existing and new alike).

trallard commented on docs/docs/community/index.mdx:
```diff
@@ -111,6 +111,10 @@ Our contributing and community guidelines take inspiration from the following pr
 - [Julia Language](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md)
 - [nteract](https://github.com/nteract/nteract/blob/main/CONTRIBUTING.md)
 
+Our gitflow process takes inspiration from the following projects:
+
+- [Bokeh](https://github.com/bokeh/bokeh/wiki/BEP-6:-Branching-Strategy)
+
```
My previous comment might have been confusing - The format here is correct, but this should be in the `Release and branching strategy` doc not in the community index 😓  apologies for the confusion

pavithraes commented:
@iameskild Thank you for raining these great points! I'm on board with your original proposal involving the `develop` branch. :)

Bokeh's strategy does indeed add a couple of extra hoops for new contributors, and to combat that we've tried to make the process clear in our contributing guide. However, Bokeh also does significantly fewer releases (definitely not one per month!) which makes this much more manageable.

iameskild commented:
Hey @pavithraes and @trallard, I have updated these docs based on the discussion we had last week around switching from Bokeh-style gitflow to Gitlab-style gitflow. This is ready for a final review. Once this is merged, I will make the requisite changes to the `nebari` repo (i.e. set default branch to `develop`).

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+---
+id: release-process-branching-strategy
```
since I cannot leave comments on files names - this needs to be under `docs/community/maintainers`

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+
+|        Cadence        |               Versioning System                |                                                                                    Release Checklist                                                                                     |                                                                                        Testing Checklist                                  ...
+| :-------------------: | :--------------------------------------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :---------------------------------------------------------------------------------------------------------------------------------------- ...
+| 3rd week of the month | [CalVer](#calver-details) - (i.e. `2022.10.1`) | [Link to issue template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=release-checklist.md&title=%5BRELEASE%5D+%3Cversion%3E) | [Link to issue template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=testing- ...
```
```suggestion
| 3rd week of the month | [CalVer](#calver-details) - (for example `2022.10.1`) | [Link to issue template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=release-checklist.md&title=%5BRELEASE%5D+%3Cversion%3E) | [Link to issue template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=testing-checklist.md&title=Testing+checklist+for+%3Cversion%3E) |
```

avoid abbreviations

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+- Communicate the status of the release to the rest of the Nebari development team and the community (through updating the checklists and adding status updates as comments).
+- Assign owners to checklist items (if not owned by the Release Captain).
+- Adjust the schedule, particularly the publishing dates, based on defects found, fixes made, holidays, vacations, and so on.
+- Ensure a new branch and milestone is created after completing a release cycle.
```
I believe this statement is no longer applicable since we are using `develop` now - so there is no need to create a new branch only the milestone

iameskild replied:
This is intermediate release branch is part of the gitlab gitflow and is what I proposed in the [original notion doc](https://www.notion.so/quansightlabs/GitFlow-Workflow-8596ceff436740e8adad9fddb328848b). 

There are two situations where this release branch comes in handy. In both instances, the `develop` branch has commits that shouldn't be included in the next release. First and most clear example is for hotfixes. Second is when there are commits made to `develop` that are not to be included in this upcoming release but rather the next release. 

`develop` branch commit history
```
commit 3 - for the dec release 
commit 2 - for the jan release 
commit 1 - for the dec release
```

Without an intermediate branch (say `release/2022.12.1`), all three of these commits are going to make into the December release which may not be what we want. With this intermediate branch, we can cherry-pick `commit 1` and `commit 3` from `develop` and cut the release with only the commits we need. 

There is an alternative and that would be to hold off merging PRs until the current release has been cut. This alternative unfortunately relies on everyone knowing which PRs belong to which release (and these things may change).

iameskild replied:
Ignore this, I misunderstood the line this was referring to. We're in agreement :sweat_smile:

trallard commented on docs/docs/community/release-process.md:
````diff
@@ -0,0 +1,68 @@
+```
+
+:::info
+`YYYY` represents the year - i.e. `2023`
+
+`MM` represents the _non-zero padded_ month - i.e. `1` for January, `12` for December
+
+`releaseNumber` represents the current release for that month, starting at `1`. Anything above `1` represents a [hotfix (patch) release](#hotfixes).
````
```suggestion
`YYYY` represents the year - such as `2023`

`MM` represents the _non-zero padded_ month - such as `1` for January, `12` for December

`releaseNumber` represents the current release for that month, starting at `1`. Anything above `1` represents a [hotfix (patch) release](#hotfixes).
```

as per our [style guide](https://www.nebari.dev/docs/community/style-guide/)

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+
+For example, the first Nebari CalVer release was `2022.10.1`. If a hotfix release was needed in the same month, we increment the `releaseNumber` by 1, which would be `2022.10.2` (_this is to illustrate how the increment works, this release does not exist._)
+
+## Gitflow and branching strategy
```
This isn't very clear - as Gitflow is a branching/contribution strategy and we are not using Gitflow (well not strict gitflow anyway)
```suggestion
## Branching strategy
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+
+## Gitflow and branching strategy
+
+Gitflow is framework for managing `git` branches by assigning certain roles to particular branches.
```
```suggestion
We use the following guidelines to manage `git` branches by assigning certain roles to particular branches.
```

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+
+Gitflow is framework for managing `git` branches by assigning certain roles to particular branches.
+
+- `develop` - Represents the active development branch and is the _default_ branch on the GitHub repository.
+
+- `main` - Represents a production-ready state of the code-base, with an appropriate tag to match the most recent release.
```
Suggestion: add a link to the GitHub references for this

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+
+- `main` - Represents a production-ready state of the code-base, with an appropriate tag to match the most recent release.
+
+- `release/YYYY-MM-releaseNumber` - Represents the branch for the upcoming release.
```
Perhaps it's worth adding a note that this branch is _only_ existent during a release cycle i.e. while actively working on making a nebari release

trallard commented on docs/docs/community/release-process.md:
```diff
@@ -0,0 +1,68 @@
+Although this process is captured in the [release checklist template](https://github.com/nebari-dev/nebari/issues/new?assignees=&labels=type%3A+release+%F0%9F%8F%B7&template=release-checklist.md&title=%5BRELEASE%5D+%3Cversion%3E), it's worth making clear how branches are managed.
+
+- Active development occurs against the `develop` branch.
+- When it's time for a release, the Release Captain will create the release branch `release/YYYY-MM-releaseNumber` and preparing the branch for the release. At times, this might mean cherry-picking a commits that are needed for this release and at other times, this might mean merging `develop` into this release branch.
```
Add a note: the release captain must ensure there are no stale release branches, and if there are, they should delete them before creating a new release branch

pavithraes replied:
```suggestion
- When it's time for a release, the Release Captain will create the release branch `release/YYYY-MM-releaseNumber` and prepare the branch for the release. At times, this might mean cherry-picking commits that are needed for this release and at other times, this might mean merging `develop` into this release branch.
```

trallard commented on docs/sidebars.js:
```diff
@@ -108,6 +108,7 @@ module.exports = {
         "community/nebari-tests",
         "community/doc-contributions",
         "community/style-guide",
+        "community/release-process-branching-strategy",
```
same comment as above - this should be on `community/maintainers`

trallard reviewed (changes requested):
There are some comments and changes that need to be addressed before merging this @iameskild 

It seems your branch is now out of sync with `main` - so you will need to bring these in-sync and move the release guide to the corresponding maintainers section

pavithraes reviewed (approved):
Thank you for working on this, @iameskild!

My only major suggestion -- I think we can move this page to the "Community/Maintainers" section, which might be a better fit. :)

The content looks complete though, so I'll approve and you can self-merge after!

iameskild commented:
@trallard @pavithraes thanks for the feedback! I think this is finally ready :)

trallard reviewed (approved):
Thanks @iameskild  this looks great. Will go ahead and merge.
